### ISSUE TITLE
In Rust 2018, forbid bulid scripts which don't print `rerun-if-chaged`

### ISSUE THREAD

**Issue by matklad** [rust-lang/cargo#5260]
The build scripts can use `rerun-if-changed` and `rerun-if-env-changed` variables to control when it's considered out-of-date. 

The problem is that if a build script does not use either of these variables, it is rerun if *anything* is changed. And this is known to cause hard to diagnose rebuilds. This is a historical behavior, so we can't change it. 

However, for Rust 2018 it makes sense to switch to one of the following behaviors:

* error if `build.rs` didn't print `rerun-if-changed`
* by default, assume `rerun-if-changed=build.rs`. That is, rerun only if build script itself changes.

**Comment by sfackler** [comment]
I think option 2 would make more sense. You clearly need to rebuild the build script if the build script changes, but in the 2018 edition, we can assume that's the only file we care about unless the build script tells us otherwise.

**Comment by matklad** [comment]
Agree, however, the potential problem here is that might become a surprising silent change if you upgrade from 2015 to 2018.

**Comment by sfackler** [comment]
Yeah that is true. You'd have to opt-in via your Cargo.toml to the edition I guess, but you may not be aware that this change is part of that?

I wonder if there are any other changes we want to make to build scripts for the edition that would make them break more obviously at the same time.

**Comment by ExpHP** [comment]
> I wonder if there are any other changes we want to make to build scripts for the edition that would make them break more obviously at the same time.

Add another level of namespacing to the flags to so that cargo can add new flags without clobbering user defined ones

```
# old way
cargo:rustc-link-lib=static=foo
cargo:rustc-link-search=native=/path/to/foo
cargo:rustc-cfg=foo
cargo:rustc-env=FOO=bar
cargo:root=/path/to/foo
cargo:libdir=/path/to/foo/lib
cargo:include=/path/to/foo/include

# new way
cargo:rustc:link-lib=static=foo
cargo:rustc:link-search=native=/path/to/foo
cargo:rustc:cfg=foo
cargo:rustc:env=FOO=bar
cargo:self:root=/path/to/foo
cargo:self:libdir=/path/to/foo/lib
cargo:self:include=/path/to/foo/include
```

:duck:

**Comment by tomjakubowski** [comment]
> Instead, what we really need to do is to add a flag to cargo build, which explains why Cargo rebuilds stuff. In particular, we can be pretty vocal about "rerunning build script because no rerun-if-changed is specified".

Was this implemented? It would be great for debugging spurious rebuilds.

**Comment by matklad** [comment]
@tomjakubowski AFAIK, not really, you might have some luck running cargo with `CARGO_LOCK=info` and looking for these log messages: https://github.com/rust-lang/cargo/blob/b8448d7c4177a27a4b14201b208ed7d2b61f39df/src/cargo/core/compiler/fingerprint.rs#L1045-L1049
